### PR TITLE
OP-1148: upgrade to Spring Boot 3.2.4

### PR DESCRIPTION
**Pull request by dbmalkovsky**
See OP-1148

Still need to work on the GUI side

### CODE REVIEW

File: .mvn/wrapper/maven-wrapper.properties
Comment (mwithi) [inline .mvn/wrapper/maven-wrapper.properties:1]: shouln't .properties be excluded from license plugin?

File: pom.xml [hunk 1]
```diff
@@ -10,43 +10,161 @@
...
+
 	<build>
 		<plugins>
 			<plugin>
 				<artifactId>maven-compiler-plugin</artifactId>
-				<version>3.12.1</version>
-				<configuration>
-					<release>${java.version}</release>
-				</configuration>
+				<version>3.11.0</version>
```
Comment (mwithi) [inline pom.xml:159]: was this supposed to be downgraded?

File: pom.xml [hunk 2]
```diff
@@ -10,43 +10,161 @@
...
-			  <configuration>
-			    <redirectTestOutputToFile>true</redirectTestOutputToFile>
-			    <reportsDirectory>${project.build.directory}/test-reports</reportsDirectory>
-			  </configuration>
+				<artifactId>maven-surefire-plugin</artifactId>
+				<version>3.2.2</version>
+				<configuration>
+					<redirectTestOutputToFile>true</redirectTestOutputToFile>
+					<reportsDirectory>${project.build.directory}/test-reports</reportsDirectory>
+				</configuration>
```
Comment (mwithi) [inline pom.xml:167]: was this supposed to be downgraded?

File: pom.xml [hunk 3]
```diff
@@ -160,7 +277,9 @@
 						<project.inceptionYear>2006</project.inceptionYear>
 						<project.name>Open Hospital</project.name>
 						<project.url>www.open-hospital.org</project.url>
-						<project.description>Open Hospital is a free and open source software for healthcare data management.</project.description>
+						<project.description>Open Hospital is a free and open source software for healthcare data
+							management.
+						</project.description>
```
Comment (mwithi) [inline pom.xml:282]: maybe unwanted newline

File: src/main/java/org/isf/medicalinventory/model/MedicalInventory.java [hunk 4]
```diff
@@ -148,4 +148,12 @@ public String getWard() {
...
 		this.ward = ward;
 	}
+
+	public int getLock() {
+		return lock;
+	}
+
+	public void setLock(int lock) {
+		this.lock = lock;
+	}
```
Comment (mwithi) [inline src/main/java/org/isf/medicalinventory/model/MedicalInventory.java:158]: This should be already landed on core, right?

File: src/main/java/org/isf/patient/service/PatientIoOperationRepository.java [hunk 5]
```diff
@@ -46,7 +46,7 @@ public interface PatientIoOperationRepository extends JpaRepository<Patient, Int
 	@Query("select p from Patient p where p.code = :id and (p.deleted = :deletedStatus or p.deleted is null)")
 	List<Patient> findAllWhereIdAndDeleted(@Param("id") Integer id, @Param("deletedStatus") char deletedStatus);
 
-	@Modifying
+	@Modifying(clearAutomatically = true)
```
Comment (mwithi) [inline src/main/java/org/isf/patient/service/PatientIoOperationRepository.java:49]: Why this was necessary here?

File: src/main/java/org/isf/menu/service/UserMenuItemIoOperationRepository.java [hunk 6]
```diff
@@ -36,7 +36,7 @@ public interface UserMenuItemIoOperationRepository extends JpaRepository<UserMen
 			"menuItem.mySubmenu, menuItem.myClass, menuItem.isASubMenu, menuItem.position, groupMenu.active " +
 			"from UserMenuItem menuItem, GroupMenu groupMenu, UserGroup  userGroup, User user " +
 			"where (user.userName=:userId) " +
-			"and (user.userGroupName=userGroup.code) " +
+			"and (user.userGroupName.code=userGroup.code) " +
```
Comment (mwithi) [inline src/main/java/org/isf/menu/service/UserMenuItemIoOperationRepository.java:39]: wondering how it was working before :/

File: src/main/java/org/isf/telemetry/envdatacollector/collectors/SoftwareDataCollector.java [hunk 7]
```diff
@@ -78,8 +78,8 @@ public Map<String, String> retrieveData() throws OHException {
 			result.put(CollectorsConstants.OS_BITNESS, String.valueOf(os.getBitness()));
 			result.put(CollectorsConstants.OS_CODENAME, os.getVersionInfo().getCodeName());
 
-			SessionImplementor sessionImp = (SessionImplementor) em.getDelegate();
-			DatabaseMetaData dbmd = sessionImp.connection().getMetaData();
+			Connection connection = em.unwrap(Connection.class);
+			DatabaseMetaData dbmd = connection.getMetaData();
```
Comment (mwithi) [inline src/main/java/org/isf/telemetry/envdatacollector/collectors/SoftwareDataCollector.java:82]: Why was this change needed?

File: src/main/java/org/isf/utils/db/DbJpaUtil.java
Comment (mwithi) [inline src/main/java/org/isf/utils/db/DbJpaUtil.java:1]: This class is still needed because of openhospital-gui/src/main/java/org/isf/lab/gui/LabBrowser.java

I think we will need an issue to remove it and to find a better way

File: src/main/resources/database.properties.sample
Comment (mwithi) [inline src/main/resources/database.properties.sample:1]: this was needed as sample in case one wants to configure an application using OH-core.jar as lib

File: src/test/resources/version.properties [hunk 8]
```diff
@@ -1,4 +1,4 @@
 # This file is not meant to be modified by the user
 VER_MAJOR=1
-VER_MINOR=13
+VER_MINOR=14
 VER_RELEASE=0
```
Comment (mwithi) [inline src/test/resources/version.properties:4]:
```suggestion
VER_RELEASE=1
```

File: src/main/java/org/isf/medicalinventory/model/MedicalInventory.java (hunk 4 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/medicalinventory/model/MedicalInventory.java:158]: No; this is part of https://github.com/informatici/openhospital-core/pull/1300

File: src/main/java/org/isf/menu/service/UserMenuItemIoOperationRepository.java (hunk 6 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/menu/service/UserMenuItemIoOperationRepository.java:39]: Same question I had; there are others like this too (including some of the tests worked when they shouldn't have)

File: src/main/java/org/isf/sms/manager/SmsManager.java [hunk 9]
```diff
@@ -102,6 +102,7 @@ else if (textLenght > MAX_LENGHT && split) {
 			for (String part : parts) {
 				Sms sms = new Sms();
 				sms.setSmsNumber(number);
+				sms.setSmsDate(smsToSend.getSmsDate());
```
[the remark this answers is not in the document]
Comment (dbmalkovsky) [reply src/main/java/org/isf/sms/manager/SmsManager.java:105]: The date has to be set because the object marks the field as NON-NULL in the column definition.   This is one of the cases that I mentioned that the newer spring boot and its dependencies are more strict.

File: src/main/java/org/isf/telemetry/envdatacollector/collectors/SoftwareDataCollector.java (hunk 7 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/telemetry/envdatacollector/collectors/SoftwareDataCollector.java:82]: Because `getDelete()` no longer exists.

File: src/test/resources/version.properties (hunk 8 above)
Comment (dbmalkovsky) [reply src/test/resources/version.properties]: Done

File: .mvn/wrapper/maven-wrapper.properties
Comment (dbmalkovsky) [reply .mvn/wrapper/maven-wrapper.properties:1]: It is; see line 267 of pom.xml.  The license in this file is the Apache license which comes with the file.  See for example,
https://github.com/jhipster/jhipster-sample-app/blob/main/.mvn/wrapper/maven-wrapper.properties

File: pom.xml (hunk 1 above)
Comment (dbmalkovsky) [reply pom.xml:159]: No; fixed

File: pom.xml (hunk 3 above)
Comment (dbmalkovsky) [reply pom.xml:282]: yes; fixed

File: src/main/java/org/isf/utils/db/DbJpaUtil.java
Comment (dbmalkovsky) [reply src/main/java/org/isf/utils/db/DbJpaUtil.java:1]: I don't see that class used any place in the GUI repo.

File: src/main/java/org/isf/patient/service/PatientIoOperationRepository.java (hunk 5 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/patient/service/PatientIoOperationRepository.java:49]: Because without that 4 tests fail.

In TestMergePatient.java:

- testMgrMergePatientPatient1MissingInformation()
- testMgrMergePatient()
- testMergePatientHistory()

And in patient/Tests.java:

- testMergePatientHistory()

File: src/main/resources/database.properties.sample
Comment (dbmalkovsky) [reply src/main/resources/database.properties.sample:1]: OK; it is put back

File: src/main/java/org/isf/utils/db/DbJpaUtil.java
Comment (mwithi) [reply src/main/java/org/isf/utils/db/DbJpaUtil.java:1]: The chain is DbJpaUtil <- DbSingleJpaConn <- PrintLabels <- LabBrowser

File: src/main/java/org/isf/sms/manager/SmsManager.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/sms/manager/SmsManager.java:105]: I see... actually the field has a default value `SMS_DATE TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP`, it would be nice to ask Spring authors how to leverage SQL defaults otherwise are they useless?

File: src/main/java/org/isf/patient/service/PatientIoOperationRepository.java (hunk 5 above)
Comment (mwithi) [reply src/main/java/org/isf/patient/service/PatientIoOperationRepository.java:49]: I imagined that, but why they fail? maybe is there an underlying design problem?

File: .mvn/wrapper/maven-wrapper.properties
Comment (mwithi) [reply .mvn/wrapper/maven-wrapper.properties:1]: oh, alright, sorry

File: src/main/java/org/isf/menu/service/UserMenuItemIoOperationRepository.java (hunk 6 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/menu/service/UserMenuItemIoOperationRepository.java:39]: No; those lines are correct as they are joining on `string`s.   Line 39 was trying to join on the variable `userGroupName which is an object  of type `UserGroup` with a `string` code value in the `userGroup` object.   Not sure how it was ever working but then the newer Hibernate and Spring JPA seem to be more strict.

File: src/main/resources/application.yml
[the remark this answers is not in the document]
Comment (dbmalkovsky) [reply src/main/resources/application.yml:1]: Application context is not part of Spring Boot 3.  Rather application.properties or application.yml are used and the rest of the information is found by reflected.

File: src/main/java/org/isf/patient/service/PatientIoOperationRepository.java (hunk 5 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/patient/service/PatientIoOperationRepository.java:49]: Maybe; it was low on my list of things to investigate deeply.  I just wanted to get the tests to work :smirk:

File: src/main/java/org/isf/sms/manager/SmsManager.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dbmalkovsky) [reply src/main/java/org/isf/sms/manager/SmsManager.java:105]: It might not be Spring; it may be Hibernate.  I am just not sure.

File: src/main/java/org/isf/utils/db/DbJpaUtil.java
Comment (dbmalkovsky) [reply src/main/java/org/isf/utils/db/DbJpaUtil.java:1]: The class still exists; I just removed methods that were not being used any longer.

**Comment by dbmalkovsky** [comment]
Note:  the changing of the `GenerationType` was required as the code was no choosing SEQUENCE instead of IDENTITY.  See https://blog.stackademic.com/avoid-using-generationtype-auto-strategy-for-id-generation-in-spring-boot-a67c6239eb7d for discussion of why using AUTO is not recommended.

**Comment by dbmalkovsky** [comment]
Note: in `UserMenuItem` the `@Convert` annotation was added because this is the only boolean column in the database that is encoded with `Y` and `N` instead of the standard `0` and `1` values.   See https://www.baeldung.com/java-hibernate-6-boolean-converters for a discussion about this change.

File: pom.xml [hunk 10]
```diff
@@ -10,43 +10,196 @@
...
+		<groupId>org.springframework.boot</groupId>
+		<artifactId>spring-boot-starter-parent</artifactId>
+		<version>3.2.4</version>
+		<relativePath/> <!-- lookup parent from repository -->
+	</parent>
+
 	<groupId>org.isf</groupId>
 	<artifactId>OH-core</artifactId>
+	<packaging>jar</packaging>
 	<version>1.14.0-SNAPSHOT</version>
```
Comment (mwithi) [inline pom.xml:23]: strange, this is supposed to be 1.14.1-SNAPSHOT now

File: src/main/resources/application.yml
[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/resources/application.yml:1]: I see

File: src/main/java/org/isf/sms/manager/SmsManager.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/sms/manager/SmsManager.java:105]: good point. what about using annotations like `@Column(nullable = false, columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP")` ?

File: pom.xml (hunk 10 above)
Comment (dbmalkovsky) [reply pom.xml]: done

File: src/main/java/org/isf/patient/service/PatientIoOperationRepository.java (hunk 5 above)
Comment (mwithi) [reply src/main/java/org/isf/patient/service/PatientIoOperationRepository.java:49]: I'm trying to understand why, despite all those tests are using the same `updateDeleted` method, they are not isolated each other and sharing _apparently_ the same persistence context while, for instance, `testMgrMergePatientMergeNotes` seems working regardless the new annotation. Are the failing methods calling each other in a way I'm not able to see?

File: src/main/java/org/isf/sms/manager/SmsManager.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dbmalkovsky) [reply src/main/java/org/isf/sms/manager/SmsManager.java]: I have found another solution.   The  "problem" is that we have an existing database; those annotations are used when generating or regenerating tables.   The suggested change isn't pretty but it is the best one I could find on the web that works with our setup.

File: src/main/java/org/isf/patient/service/PatientIoOperationRepository.java (hunk 5 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/patient/service/PatientIoOperationRepository.java:49]: See https://www.baeldung.com/spring-data-jpa-modifying-annotation and https://stackoverflow.com/questions/43665090/why-do-we-have-to-use-modifying-annotation-for-queries-in-data-jpa for the reasons.   I think we should change it to `@Modifying(clearAutomatically=true, flushAutomatically=true)`

File: src/main/java/org/isf/sms/manager/SmsManager.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/sms/manager/SmsManager.java]: I remember this kind of solution, it's ok for me, tested?

File: src/main/java/org/isf/patient/service/PatientIoOperationRepository.java (hunk 5 above)
Comment (mwithi) [reply src/main/java/org/isf/patient/service/PatientIoOperationRepository.java:49]: it looks like we are facing this "problem" now because of the introduction of `org.springframework.boot.test.context.SpringBootTest` in the new `OHCoreTestCase.java`, which anyway might be a symptom that tests are not correctly written causing uncommited transactions or uncompleted rollbacks (I guess the latter). If the uncommitted transactions and uncompleted rollbacks are exactly what we are testing, then the "problem" is expected and using clearAutomatically and flushAutomatically sounds reasonable, otherwise we should understande the root cause of the fails.

File: src/main/java/org/isf/sms/manager/SmsManager.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dbmalkovsky) [reply src/main/java/org/isf/sms/manager/SmsManager.java]: yes; the tests all pass

File: src/main/java/org/isf/patient/service/PatientIoOperationRepository.java (hunk 5 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/patient/service/PatientIoOperationRepository.java:49]: IHMO, all the tests need to be reviewed.   Some of them are confusing and convoluted at the very least.   I think we can do that when I rename the test files (discussed previously) and create a new "draft" PR for  a newer and more complete testing structure.

**Review by mwithi — APPROVED** [review]
LGTM. Thanks a lot!
